File: R/example17_2.R
# Example 17.2: Ex-Ante Forecasts
# Using package forecast

setwd("C:/Course17/ceR/R")
# read data from 1959.1 to 2003.4 (181 obs.)
z<-read.table("http://web.pdx.edu/~crkl/ceR/data/gdp96.txt",header=T)
gdp<-ts(z$GDP,start=c(1959,1),frequency=4)
pgdp<-ts(z$PGDP2000,start=c(1959,1),frequency=4)
# constant scenario (0% AGR)
leading<-ts(c(z$LEADING96,115.0,115.0,115.0,115.0,115.0),start=c(1959,1),frequency=4)
# pessimistic scenario (-2% AGR)
# leading<-ts(c(z$LEADING96,115.0,114.4,113.8,113.2,112.6),start=c(1959,1),frequency=4)
# constant scenario (0% AGR)
# leading<-ts(c(z$LEADING96,115.0,115.0,115.0,115.0,115.0),start=c(1959,1),frequency=4)
# optimistic scenario (+2% AGR)
# leading<-ts(c(z$LEADING96,115.0,115.6,116.2,116.8,117.4),start=c(1959,1),frequency=4)
rgdp<-100*gdp/pgdp
growth<-100*(rgdp-lag(rgdp,-4))/lag(rgdp,-4)
xvar<-ts.intersect(lag(leading,-1),lag(leading,-5))  

growth.train<-window(growth,start=c(1961,1),end=c(2001,4))
xvar.train<-window(xvar,start=c(1961,1),end=c(2001,4))
growth.test<-window(growth,start=c(2002,1))
xvar.test<-window(xvar,start=c(2002,1))

arma1<-arima(growth.train,order=c(1,0,4),xreg=xvar.train)
arma1
for1<-predict(arma1,n.ahead=14,newxreg=xvar.test)
for1
plot(for1$pred)
lines(growth.test,lty=2)
lines(for1$pred+for1$se,col="red")
lines(for1$pred-for1$se,col="red")

library(forecast)
arma2<-Arima(growth.train,order=c(1,0,4),xreg=xvar.train)
arma2
for2<-forecast(arma2,h=14,xreg=xvar.test)
for2
accuracy(for2,growth.test)
plot(for2)
plot(for2,include=10)
lines(growth.test)

File: python/example17_2.py
# Example 17.2: Ex-Ante Forecasts

import numpy as np
import pandas as pd

# read data from 1959.1 to 2003.4 (181 obs.)
z = pd.read_csv('http://web.pdx.edu/~crkl/ceR/data/gdp96.txt',sep='\s+')
# setup time series index
Yr = (10*z['QUARTER'])//10
Qt = (10*z['QUARTER'])%10
z.index = pd.PeriodIndex(year=Yr,quarter=Qt,freq='Q-DEC')
# z.index = pd.period_range('1959Q1', '2003Q4', freq='Q-DEC')
# ex-ante forecast period and assumption
fidx = pd.period_range('2004Q1','2005Q1',freq='Q-DEC')
# constant scenario (0% AGR)
f1 = pd.DataFrame(data=[115.0,115.0,115.0,115.0,115.0],index=fidx)
# pessimistic scenario (-2% AGR)
# f1 = pd.DataFrame(data=[115.0,114.4,113.8,113.2,112.6],index=fidx)
# optimistic scenario (+2% AGR)
# f1 = pd.DataFrame(data=[115.0,115.6,116.2,116.8,117.4],index=fidx)

rgdp = 100*z['GDP']/z['PGDP2000']
growth = 100*(rgdp/rgdp.shift(4)-1.0)
leading = pd.concat([z['LEADING96'],f1])
xvar = pd.concat([leading.shift(1),leading.shift(5)],axis=1)
xvar.columns = ['LEADING-1','LEADING-5']
growth_train = growth['1961Q1':'2001Q4']
xvar_train = xvar['1961Q1':'2001Q4']
growth_test = growth['2002Q1':]
xvar_test = xvar['2002Q1':]

from statsmodels.tsa.arima_model import ARMA
arma1 = ARMA(growth_train,exog=xvar_train,order=(1,4)).fit()
print(arma1.summary())

import matplotlib.pyplot as plt 
%matplotlib inline
for1 = arma1.predict(start='2002Q1',end='2005Q1',exog=xvar_test)
for1.plot()
plt.plot(growth_test)
plt.show()
print(for1)

for2 = arma1.forecast(steps=13,exog=xvar_test,alpha=0.05)
for2[0]  # forecast value
for2[1]  # forecast se
for2[2]  # forecast interval
for2data ={'actural':growth_test, 
           'forecast':for2[0],
           's.e.':for2[1],
           'lower':for2[2][:,0],
           'upper':for2[2][:,1]}
for2index = pd.period_range(start='2002Q1',end='2005Q1',freq='Q-DEC')
forecast = pd.DataFrame(data=for2data,index=for2index)
forecast.plot(y=['actural','forecast','lower','upper'])
print(forecast)
